### PR TITLE
feat(iota-e2e-tests): Include object snapshot generation and address swap in migration test

### LINKED ISSUE #4379: [genesis-builder] Create and e2e test for the Address Swap Map

### PR DESCRIPTION
**Pull request by miker83z**
# Description of change

This PR increases the operations executed in the e2e test ´test_full_node_load_migration_data´ by 
- adding the generation of a Stardust Object Snapshot coming from an (IOTA) Hornet full snapshot 
- performing an address swap in the migration (the previous step was necessary for performing this).

A refactoring was needed in the `iota-genesis-builder` main, in order to make available some Hornet output filter functionalities to other crates. 

## Links to any relevant issues

Fixes #4379 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

- `cargo test --package iota-e2e-tests --test full_node_migration_tests -- test_full_node_load_migration_data --exact --show-output`
- `cargo run --release --bin iota-genesis-builder -- --disable-global-snapshot-verification iota --snapshot-path crates/iota-e2e-tests/tests/migration/test_hornet_full_snapshot.bin --target-network alphanet-test --address-swap-map-path crates/iota-e2e-tests/tests/migration/address_swap.csv`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Dkwcs — APPROVED** [review]

File: crates/iota-e2e-tests/tests/full_node_migration_tests.rs [hunk 1]
```diff
@@ -39,26 +54,67 @@ const SPONSOR_ADDRESS_MNEMONIC: &str = "okay pottery arch air egg very cave cash
...
-
     let IotaTransactionBlockResponse {
         confirmed_local_execution,
         errors,
         ..
     } = tx_response;
+
+    // The transaction must be successful
     assert!(confirmed_local_execution.unwrap());
     assert!(errors.is_empty());
```
Comment (valeriyr) [inline crates/iota-e2e-tests/tests/full_node_migration_tests.rs:88]: Should we add a check that the address swap is applied?

**Review by valeriyr — APPROVED** [review]

File: crates/iota-e2e-tests/tests/full_node_migration_tests.rs (hunk 1 above)
Comment (miker83z) [reply crates/iota-e2e-tests/tests/full_node_migration_tests.rs:88]: This check is part of the migration run [here](https://github.com/iotaledger/iota/pull/4314/files#diff-cd61891c45eb4425642dc5d82826657f701090ed1af8d358d61f2a1c39889d65R141). So if the migration execution is correct, then the address swap has happened.

File: crates/iota-e2e-tests/tests/full_node_migration_tests.rs [hunk 2]
```diff
@@ -39,26 +55,67 @@ const SPONSOR_ADDRESS_MNEMONIC: &str = "okay pottery arch air egg very cave cash
...
+    // snapshot
+    let dir = tempdir()?;
+    let startudst_object_snapshot_file_path = dir.path().join(MIGRATION_DATA_FILE_NAME);
+    let object_snapshot_writer =
+        BufWriter::new(File::create(&startudst_object_snapshot_file_path)?);
+
+    // Generate the stardust object snapshot
+    genesis_builder_snapshot_generation(object_snapshot_writer)?;
+    // Then load it
+    let snapshot_source = SnapshotSource::Local(startudst_object_snapshot_file_path);
```
Comment (DaughterOfMars) [inline crates/iota-e2e-tests/tests/full_node_migration_tests.rs]: typo here `startudst_object_snapshot_file_path`

Comment (miker83z) [reply crates/iota-e2e-tests/tests/full_node_migration_tests.rs]: Solved, thanks!

**Review by kodemartin — APPROVED** [review]
lgtm :bird:

**Pull request merged**